src/ray/pubsub/publisher.cc

```diff
@@ -20,149 +20,183 @@ namespace pubsub {
+                                              const SubscriberID &subscriber_id) {
   // Erase from subscribers_to_objects_;
-  auto subscribers_to_objects_it = subscribers_to_objects_.find(subscriber_id);
-  if (subscribers_to_objects_it == subscribers_to_objects_.end()) {
+  const auto key_id = KeyIdType::FromBinary(key_id_binary);
+  auto subscribers_to_message_it = subscribers_to_key_id_.find(subscriber_id);
+  if (subscribers_to_message_it == subscribers_to_key_id_.end()) {
     return false;
   }
-  auto &objects = subscribers_to_objects_it->second;
-  auto object_it = objects.find(object_id);
+  auto &objects = subscribers_to_message_it->second;
+  auto object_it = objects.find(key_id);
   if (object_it == objects.end()) {
-    RAY_CHECK(objects_to_subscribers_.count(object_id) == 0);
+    RAY_CHECK(key_id_to_subscribers_.count(key_id) == 0);
     return false;
   }
   objects.erase(object_it);
   if (objects.size() == 0) {
-    subscribers_to_objects_.erase(subscribers_to_objects_it);
+    subscribers_to_key_id_.erase(subscribers_to_message_it);
   }
 
   // Erase from objects_to_subscribers_.
-  auto objects_to_subscribers_it = objects_to_subscribers_.find(object_id);
+  auto key_id_to_subscriber_it = key_id_to_subscribers_.find(key_id);
   // If code reaches this line, that means the object id was in the index.
-  RAY_CHECK(objects_to_subscribers_it != objects_to_subscribers_.end());
-  auto &subscribers = objects_to_subscribers_it->second;
+  RAY_CHECK(key_id_to_subscriber_it != key_id_to_subscribers_.end());
+  auto &subscribers = key_id_to_subscriber_it->second;
   auto subscriber_it = subscribers.find(subscriber_id);
   // If code reaches this line, that means the subscriber id was in the index.
   RAY_CHECK(subscriber_it != subscribers.end());
   subscribers.erase(subscriber_it);
   if (subscribers.size() == 0) {
-    objects_to_subscribers_.erase(objects_to_subscribers_it);
+    key_id_to_subscribers_.erase(key_id_to_subscriber_it);
   }
   return true;
 }
 
-bool SubscriptionIndex::AssertNoLeak() const {
-  return objects_to_subscribers_.size() == 0 && subscribers_to_objects_.size() == 0;
+template <typename KeyIdType>
+bool SubscriptionIndex<KeyIdType>::CheckNoLeaks() const {
+  return key_id_to_subscribers_.size() == 0 && subscribers_to_key_id_.size() == 0;
 }
 
-bool Subscriber::ConnectToSubscriber(
-    LongPollConnectCallback long_polling_reply_callback) {
-  if (long_polling_reply_callback_ == nullptr) {
-    long_polling_reply_callback_ = long_polling_reply_callback;
+bool Subscriber::ConnectToSubscriber(rpc::PubsubLongPollingReply *reply,
+                                     rpc::SendReplyCallback send_reply_callback) {
+  if (!long_polling_connection_) {
+    RAY_CHECK(reply != nullptr);
+    RAY_CHECK(send_reply_callback != nullptr);
+    long_polling_connection_ =
+        absl::make_unique<LongPollConnection>(reply, std::move(send_reply_callback));
```

Comment:
Since C++14 is enabled in Ray, use `std::make_unique` instead because the type of `long_polling_connection_` is std::unique_ptr.

Reply:
Good catch. Will fix.